text/0017-parameter_type_with_encoding.md

```diff
@@ -0,0 +1,143 @@
+This problem cannot be fixed by protocol bits, and is in particular difficult when multiple GCS are based off a common source that might take some releases to get into sync.
+
+The proposed soluton is that a bitmask extension field be added to `PARAM_REQUEST_READ` and `PARAM_REQUEST_LIST` which flags which of the new data types the requestor understands.
+The flight stack then configures itself to send the BYTEWISE encoded value of the appropriate type if supported, and otherwise a supported fallback type. 
+
+With the current proposed set of items the flag would have to be a 32-bit bitmask - there will be 10 x 3 parameters.
+
+Flags would be `MAV_PARAM_TYPE_SUPPORTED` values like `MAV_PARAM_TYPE_SUPPORTED_INT32`, `MAV_PARAM_TYPE_SUPPORTED_CAST_INT32`, `MAV_PARAM_TYPE_SUPPORTED_BYTEWISE_INT32`
+
+
+Notes:
+- In future this set may be extended with additional flight modes and component specific modes.
+- The "return home" and "safe recovery" modes both serve the same function: getting the vehicle to a safe location in the event of a failsafe or for landing.
+  They are separated because from the user perspective the behaviour is quite different and the GCS needs to be able to present that behaviour differently.
+  - The existance of these as separate modes means that the autopilot must have separate custom modes for these if we don't publish the standard mode independently. 
+
+## Protocol bits
+
+The two protocol bits `MAV_PROTOCOL_CAPABILITY_PARAM_ENCODE_BYTEWISE` and `MAV_PROTOCOL_CAPABILITY_PARAM_ENCODE_C_CAST` will have to be updated to state that they only apply to the type flags that do not include encoding.
+
+Question 
+- what is the discoverability mechanism for parameter protocol?
+  Right now you should set one of the above bits, and a GCS can use that to determine support.
+  But in future if we all moved to using these new flags, how would we do it?
+
+
+## Deprecation Path
+
+TBD. 
+
+This update complicates the protocol.
+- It adds new enum bits - where there are 10 there may now be 30.
+  It is not as obvious which should be selected.
+- It adds mechanism to query supported types, which are used in parallel with the protocol bits.
+  - We want to make it clear which you should use and when.
+
+What is the envisaged end point? I'd like it to be simple.
+
+
+# Alternatives
+
+- Do nothing.
+  - A GCS will be able to decode messages if it reads the protocol bit (once that is set)
+  - Over time a big-bang migration to bytewise encoding would be possible if GCS update to using the protocol bits for identity.
+    - I have my doubts this could be reliably managed. Whereas if ground stations start receiving these odd new enums I think they will adjust.
+- ?
+
+# Unresolved Questions
+
+- [ ] Do we need C-style encoding enums?
+- [ ] What does the end point look like? i.e. the end protocol is more complicated.
+- [ ] Are there other alternatives to aid in migration
+- [ ] Is there any reason for PX4 to update to the new types?
```

Comment:
This is relevant. It might be that the impetus is to have a clean common set of instructions at the end of the process.